packages/qwik-city/src/middleware/request-handler/resolve-request-handlers.ts

```diff
@@ -257,9 +227,49 @@ export function actionsMiddleware(routeActions: ActionInternal[], routeLoaders:
 
         return loaders[loaderId];
       });
+    }
 
-      await Promise.all(resolvedLoadersPromises);
+    if (method === 'POST') {
+      const selectedActionId = requestEv.query.get(QACTION_KEY);
+      if (selectedActionId) {
+        const serverActionsMap = globalThis._qwikActionsMap as
+          | Map<string, ActionInternal>
+          | undefined;
+        const action =
+          routeActions.find((action) => action.__id === selectedActionId) ??
+          serverActionsMap?.get(selectedActionId);
+        if (action) {
+          requestEv.sharedMap.set(RequestEvSharedActionId, selectedActionId);
+          const data = await requestEv.parseBody();
+          if (!data || typeof data !== 'object') {
+            throw new Error(
+              `Expected request data for the action id ${selectedActionId} to be an object`
+            );
+          }
+          const result = await runValidators(requestEv, action.__validators, data, isDev);
+          if (!result.success) {
+            loaders[selectedActionId] = requestEv.fail(result.status ?? 500, result.error);
+          } else {
+            //@ts-ignore
+            const actionResolved = isDev
+              ? await measure(requestEv, action.__qrl.getSymbol().split('_', 1)[0], () =>
+                  action.__qrl.call(requestEv, result.data as JSONObject, requestEv)
+                )
+              : await action.__qrl.call(requestEv, result.data as JSONObject, requestEv);
+            if (isDev) {
+              verifySerializable(qwikSerializer, actionResolved, action.__qrl);
+            }
+
+            loaders[selectedActionId] = actionResolved;
+            isAction = true;
+          }
+        }
+      }
+    }
+    if (routeLoaders.length > 0 && isAction) {
+      await Promise.all(loader() || []);
     }
+    isAction = false;
```

Comment:
Do we need to reset it to false?

Reply:
We need to prevent it from running multiple times.